From c5e07b400da803b741addc676365d957badacadc Mon Sep 17 00:00:00 2001
From: Noam Lessner <nlessner@habana.ai>
Date: Sun, 10 Nov 2019 11:55:32 +0200
Subject: [PATCH] [SW-6739] arcboot: custom memory map and startup

Change linker memory map to custom according to our SOC.
Define custom startup to do all needed before calling main().

Change-Id: I10dfb33616a5bb0f0c28859f67bd5cda6f1b4542
---

diff --git a/Makefile b/Makefile
index 74ba1cc..94064a6 100644
--- a/Makefile
+++ b/Makefile
@@ -15,25 +15,30 @@
 CFLAGS += -Isdk/src/include
 CFLAGS += -Isdk/elpsoft/src/headers
 CFLAGS += -Isdk/elpsoft/src/math
+
 LDFLAGS = -flto
 LDFLAGS += -L $(LIBDIR) -lsb -lellipsys
-LDFLAGS += -Wl,--defsym=lf2flt=0x124
+LDFLAGS += -T arcboot.lds
+LDFLAGS += -nostartfiles
 
 .PHONY: all libsb libellipsys sbtool
-all: arcboot.bin arcboot_rom.bin
-	@touch arcboot.bin
+all: libsb libellipsys
+	@$(MAKE) arcboot.bin arcboot_rom.bin
 
 libsb libellipsys:
 	@$(MAKE) -C sdk $@ SB_SIGN_MODE=rsa_std
 sbtool:
 	@$(MAKE) -C sdk $@
 %.bin: %
-	@$(OBJCOPY) -O binary -R .note.gnu.build-id $< $@
+	@$(OBJCOPY) -O binary -R .note.gnu.build-id -R .*data -R .test* $< $@
 
-arcboot: arcboot.o libsb libellipsys
-	@$(CC) $< $(LDFLAGS) -o $@
-arcboot_rom: arcboot_rom.o libsb libellipsys
-	@$(CC) $< $(LDFLAGS) -o $@
+arcboot: arcboot.o crt0.o
+	$(CC) $^ $(LDFLAGS) -o $@
+arcboot_rom: arcboot_rom.o crt0.o
+	@$(CC) $^ $(LDFLAGS) -o $@
 
 %.o: %.c
 	@$(CC) $(CFLAGS) -c $< -o $@
+
+%.o: %.s
+	@$(CC) $(AFLAGS) -c $< -o $@
diff --git a/arcboot.lds b/arcboot.lds
new file mode 100644
index 0000000..a72cd28
--- /dev/null
+++ b/arcboot.lds
@@ -0,0 +1,88 @@
+/*
+ * (C) Copyright 2017 HabanaLabs, Ltd.
+ * Noam Camus <noamikong@gmail.com>
+ *
+ * SPDX-License-Identifier:     GPL-2.0+
+ */
+
+OUTPUT_FORMAT("elf32-littlearc", "elf32-bigarc", "elf32-littlearc")
+OUTPUT_ARCH(arc)
+
+MEMORY
+{
+	ICCM	: ORIGIN = 0x00000000, LENGTH = 64K
+	SRAM	: ORIGIN = 0x10000000, LENGTH = 128M
+	STM		: ORIGIN = 0x18000000, LENGTH = 32M
+	SPI		: ORIGIN = 0x1A000000, LENGTH = 16M
+	SBROM	: ORIGIN = 0x1B040000, LENGTH = 64K
+	FWSRAM	: ORIGIN = 0x1B050000, LENGTH = 64K
+	DCCM	: ORIGIN = 0x80000000, LENGTH = 16K
+	CS		: ORIGIN = 0xf0000000, LENGTH = 64M
+}
+
+REGION_ALIAS("REGION_TEXT",		ICCM)
+REGION_ALIAS("REGION_DATA",		FWSRAM)
+REGION_ALIAS("REGION_SDATA",	DCCM)
+REGION_ALIAS("REGION_SBSS",		DCCM)
+REGION_ALIAS("REGION_BSS",		DCCM)
+REGION_ALIAS("REGION_HEAP",		DCCM)
+REGION_ALIAS("REGION_STACK",	DCCM)
+
+ENTRY(__start)
+
+SECTIONS
+{
+	.text : {
+		crt0.o (.text*)
+		*(.text)
+	} > REGION_TEXT
+
+	.data : {
+		*(.data)
+	} > REGION_DATA
+
+	.sdata : {
+		__SDATA_BEGIN__ = .;
+		*(.sdata .sdata.* .gnu.linkonce.s.*)
+		_edata = .;
+		PROVIDE (edata = .);
+	} > REGION_SDATA
+
+	.sbss : {
+		__bss_start = .;
+		PROVIDE (__sbss_start = .);
+		PROVIDE (___sbss_start = .);
+		*(.dynsbss)
+		*(.sbss .sbss.* .gnu.linkonce.sb.*)
+		*(.scommon)
+		PROVIDE (__sbss_end = .);
+		PROVIDE (___sbss_end = .);
+	} > REGION_SBSS
+
+	.bss : {
+		*(.dynbss)
+		*(.bss .bss.* .gnu.linkonce.b.*)
+		*(COMMON)
+		/* Align here to ensure that the .bss section occupies space up to
+		_end. Align after .bss to ensure correct alignment even if the
+		.bss section disappears because there are no input sections. */
+		. = ALIGN(32 / 8);
+		_end = .;
+		PROVIDE (end = .);
+	} > REGION_BSS
+
+	.heap : {
+		. = ALIGN(32 / 8);
+		__start_heap = . ;
+		. = . + (DEFINED(__HEAP_SIZE) ? __HEAP_SIZE : (DEFINED(__DEFAULT_HEAP_SIZE) ? __DEFAULT_HEAP_SIZE : 1k)) ;
+		__end_heap = . ;
+		. = ALIGN(0x8);
+	} > REGION_HEAP
+
+	.stack : {
+		. = ALIGN(32 / 8);
+		__stack = . ;
+		. = . + (DEFINED(__STACK_SIZE) ? __STACK_SIZE : (DEFINED(__DEFAULT_STACK_SIZE) ? __DEFAULT_STACK_SIZE : 512)) ;
+		__stack_top = . ;
+	} > REGION_STACK
+}
\ No newline at end of file
diff --git a/crt0.S b/crt0.S
new file mode 100644
index 0000000..5e572a3
--- /dev/null
+++ b/crt0.S
@@ -0,0 +1,57 @@
+;; *******************************************************************************************************************
+;; -----------------
+;; ARC start up file
+;; -----------------
+;; The startup code for the ARC family of processors does the following
+;; before transferring control to user defined main label :
+;;	1. Set sp to __stack_top (link time variable)
+;;	2. Set fp to zero
+;;	3. Zero out the bss section (for uninitialized globals)
+;; After returning from main, the processor is halted and the pipeline is flushed out
+;;
+;; *******************************************************************************************************************
+.file "crt0.S"
+
+.extern main
+.text
+
+;; lf2flt: to avoid compiler (buildroot wrapper) warning during linkage
+.global	lf2flt
+lf2flt:
+
+.global __start
+.type __start, @function
+__start:
+	mov gp, @__SDATA_BEGIN__
+
+	mov_s r2, @__sbss_start	; r2 = start of the bss section
+	sub r3, @_end, r2		; r3 = size of the bss section in bytes
+	; set up the loop counter register to the size (in words) of the bss section
+	asr_s r13, r3
+	asr.f lp_count, r13
+
+	mov sp, @__stack_top	; initialize stack pointer
+	mov_s r3, 0
+	; loop to zero out the bss.  Enter loop only if lp_count != 0
+	lpnz @.Lend_zbss
+	st.ab r3, [r2, 4]		; zero out the word
+	nop
+.Lend_zbss:
+
+	;; branch to main
+	bl.d @main
+	mov fp, 0				; initialize frame pointer
+							; r0 contains exit code
+	j @_exit_halt
+
+.global _exit_halt
+.type _exit_halt, @function
+.align 4
+_exit_halt:
+	; r0 contains exit code
+	flag 0x01
+	nop
+	nop
+	nop
+	j @_exit_halt
+	nop
